src/data/moves/move-attrs/add-type-attr.ts

```diff
@@ -42,4 +49,50 @@ export class AddTypeAttr extends MoveEffectAttr {
   override getCondition(): MoveConditionFunc {
     return (_user, target, _move) => !target.isTerastallized && !target.getTypes().includes(this.type);
   }
+
+  /**
+   * @returns The inverse {@link getTypeChangeScore | score} for adding this effect's type
+   * to the given target.
+   */
+  public override getEffectScore(user: EnemyPokemon, target: Pokemon, _move: Move): number {
+    return -this.getTypeChangeScore(user, target);
+  }
+
+  /**
+   * @returns A {@link getTypeChangeScore | score} for adding this effect's type to the given target.
+   * If this score is negative, this defers to the {@linkcode ALLY_TARGET_PENALTY} instead
+   */
+  public override getAllyTargetScore(user: EnemyPokemon, target: EnemyPokemon, _move: Move): number | null {
+    const typeChangeScore = this.getTypeChangeScore(user, target);
+    return typeChangeScore >= 0 ? typeChangeScore : null;
+  }
+
+  /**
+   * Calculates a score to estimate how much the target would benefit from
+   * gaining this effect's type. A target is perceived to benefit from this effect if
+   * - The added type reduces the effectiveness of all known attacks from the target's opponents.
+   * - The user is faster than all of its active opponents.
+   *
+   * Conversely, this effect is perceived as a drawback if the added type increases the effectiveness
+   * of at least one attack from the target's opponent(s).
+   * @param user - The {@linkcode EnemyPokemon} evaluating this effect
+   * @param target - The {@linkcode Pokemon} this effect is evaluated against
+   * @returns An integer score representing how much this effect benefits the target
+   */
+  private getTypeChangeScore(user: EnemyPokemon, target: Pokemon): number {
+    const oppMoveTypes = new Set(
+      target.getOpponents().flatMap((opp) => {
+        const attackMoves = opp.isOpponent(user) ? opp.estimateAttackMoves() : opp.getAttackMoves(true);
+        return attackMoves.map((mv) => opp.getMoveType(mv));
+      }),
+    );
+
+    const modDamageMultiplier = Math.max(...[...oppMoveTypes].map((t) => getTypeDamageMultiplier(t, this.type)));
+    if (modDamageMultiplier >= 1) {
+      return modDamageMultiplier > 1 ? MINOR_EFFECT_SCORE_PENALTY : 0;
+    }
+
+    const userOutspeeds = user.getOpponents().every((opp) => user.outspeeds(opp, true));
+    return userOutspeeds || !globalScene.currentBattle.double ? MINOR_EFFECT_SCORE_BONUS : 0;
```

Comment:
```suggestion
    if (modDamageMultiplier > 1) {
      return MINOR_EFFECT_SCORE_PENALTY;
    } else if (modDamageMultiplier === 1) {
      return 0;
    } else if (!globalScene.currentBattle.double) {
      return MINOR_EFFECT_SCORE_BONUS;
    }

    const userOutspeeds = user.getOpponents().every((opp) => user.outspeeds(opp, true));
    return userOutspeeds ? MINOR_EFFECT_SCORE_BONUS : 0;
```
Since single battles bypass the outspeed check, might as well not make the check at all if it's a single battle.